compiler/ir/ir.psi2ir/src/org/jetbrains/kotlin/psi2ir/generators/ArgumentsGenerationUtils.kt

```diff
@@ -422,17 +425,23 @@ fun StatementGenerator.generateSamConversionForValueArgumentsIfRequired(call: Ca
                 "$originalDescriptor has ${originalDescriptor.typeParameters}"
     }
 
+    val partialSam = context.languageVersionSettings.supportsFeature(LanguageFeature.SamConversionPerArgument)
     val substitutionContext = call.original.typeArguments.entries.associate { (typeParameterDescriptor, typeArgument) ->
         underlyingDescriptor.typeParameters[typeParameterDescriptor.index].typeConstructor to TypeProjectionImpl(typeArgument)
     }
     val typeSubstitutor = TypeSubstitutor.create(substitutionContext)
 
     for (i in underlyingValueParameters.indices) {
-        val originalParameterType = originalValueParameters[i].type
         val underlyingParameterType = underlyingValueParameters[i].type
 
-        if (!context.extensions.samConversion.isSamType(underlyingParameterType)) continue
-        if (!originalParameterType.isFunctionTypeOrSubtype) continue
+        if (partialSam && resolvedCall is NewResolvedCallImpl<*>) {
+            // TODO support SAM conversion of varargs
+            val argument = resolvedCall.valueArguments[originalValueParameters[i]]?.arguments?.singleOrNull() ?: continue
```

Comment:
NB ideally, at some point FE should provide such information explicitly. There are some "dark corners" here (such as smart cast vs SAM conversion), currently PSI2IR makes an "educated guess" regarding what FE actually had in mind.